File: 9.贪心算法/8.跳跃游戏II/45.跳跃游戏-ii.py
#
# @lc app=leetcode.cn id=45 lang=python3
#
# [45] 跳跃游戏 II
#
from mytools import *
# @lc code=start
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        res, ends, far = 0, 0, 0
        for i in range(0, n - 1):
            far = max(far, i + nums[i])
            if i == ends:
                res += 1
                ends = far
        return res
# @lc code=end

File: 9.贪心算法/8.跳跃游戏II/45.跳跃游戏-ii.cpp
/*
 * @lc app=leetcode.cn id=45 lang=cpp
 *
 * [45] 跳跃游戏 II
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    int jump(vector<int>& nums) {
        int nxt = 0, res = 0, ends = 0;
        for (int i = 0; i < nums.size(); i ++ )
        {
            nxt = max(nxt, i + nums[i]);
            if (i == ends)
            {
                if (ends < nums.size() - 1)
                {
                    ends = nxt;
                    res ++;
                }
                if (ends >= nums.size() - 1) break;
            }
        }
        return res;
    }
};
// @lc code=end

